### PR TITLE
Add Semimodule type class

### PR DESCRIPTION
**Pull request by goverthrow**
These changes incorporate #215 in a different way to satisfy #216

### CODE REVIEW

File: plutus-numeric/src/PlutusTx/Numeric/Extra.hs [hunk 1]
```diff
@@ -410,19 +455,18 @@ infixr 8 ^
 
  @since 4.2
 -}
-{-# INLINEABLE scaleNat #-}
+{-# DEPRECATED scaleNat "Inefficient. Use 'semiscale' instead" #-}
+{-# INLINEABLE [2] scaleNat #-}
```
Comment (kozross) [inline plutus-numeric/src/PlutusTx/Numeric/Extra.hs:459]: You can drop the `[2]` on that. This is a remnant of my attempt to make rewrites work.

File: plutus-numeric/src/PlutusTx/Numeric/Laws.hs [hunk 2]
```diff
@@ -224,45 +224,45 @@ integralDomainLaws =
...
-scaleNatLaws =
-  law "scaleNat n (r1 + r2) = scaleNat n r1 + scaleNat n r2" scaleNatDist
-    <> law "scaleNat n1 (scaleNat n2 r) = scaleNat (n1 * n2) r" scaleNatComp
-    <> law "scaleNat one r = r" scaleNatOne
-    <> law "scaleNat zero r = zero" scaleNatZero
+semimoduleLaws =
+  law "semiscale x z + semiscale y z = semiscale (x + y) z" semiscaleDist
+    <> law "semiscale x (semiscale y z) = semiscale (x * y) z" semiscaleComp
+    <> law "semiscale zero x = zero" semiscaleOne
+    <> law "semiscale one x = x" semiscaleZero
```
Comment (kozross) [inline plutus-numeric/src/PlutusTx/Numeric/Laws.hs:235]: Slight naming issue?

**Review by kozross — APPROVED** [review]

**Comment by kozross** [comment]
@goverthrow - slight misclick on my part: I meant to click 'request changes'.

**Comment by goverthrow** [comment]
Fixed

**Review by kozross — APPROVED** [review]

**Pull request merged**